### ISSUE TITLE
New stub mapping record/playback feature

### ISSUE THREAD

**Issue by tomakehurst** [tomakehurst/wiremock#546]
# Problem
WireMock's record feature seems to be one of the most heavily used, and as such it is quite frequently the subject of feature requests. Rather than apply lots of spot fixes for specific cases, I propose re-writing the record feature from scratch, incorporating general solutions to the types of problems expressed here and on the mailing list.

Based on past questions and requests related to the recorder, I'd like to suggest it attempts to achieve the following things:
* Record and playback without having to restart in between.
* Customise output: file naming, request/response headers/body transformation.
* Filter what gets recorded.
* Statefulness - support a succession of responses for the same request (rather than just ignoring subsequent responses as is currently the case)
* Choose target - in memory, persisted, downloaded
* Potential to import from other formats e.g. HAR, RAML (but not necessarily immediate support)

# Solution
I propose the following solution based on these objectives:

**Snapshot recording** - make some requests (that get captured in the request journal) then hit a "capture snapshot" API. This avoids the stop/start problem. It also means that optional filtering and transformation parameters could be passed into the API call, in addition to a save/download flag.

**Use Scenarios for multiple identical requests** - when more than one instance of a request is seen, capture each under one scenario so that the progression of responses seen originally is played back.

**Recording transformer extension point** - expose a Java interface that (similar to `ResponseDefinitionTransformer`) supports custom transformations of stub mappings as they're recorded.

**Recording adapter extension point** - expose a Java API to the recorder which is sufficiently generic that other formats can be adapted to it e.g. Swagger/RAML or HAR.

**Comment by scho** [comment]
Would this solve #542 as well?

**Comment by tomakehurst** [comment]
Yes it would

**Comment by selangley** [comment]
Looks to be very compatible with the operations I proposed in my pull request last year:

https://github.com/tomakehurst/wiremock/pull/400
Added ability to record a particular request/response repeatedly

so I'm all for this solution.

**Comment by sabavatbbm** [comment]
is **Use Scenarios for multiple identical requests** already implemented? How do I use it?
Also, can we have sub directories in the __files folder?

**Comment by tomakehurst** [comment]
@sabavatbbm yes, scenarios are already implemented. What I'm suggesting here is that the recording mappings make use of this feature to step through multiple states. See: http://wiremock.org/docs/stateful-behaviour/

Yes, subdirectories are fine under `__files` (and `mappings` for that matter).

**Comment by MasonM** [comment]
[I posted this to the mailing list](https://groups.google.com/forum/#!topic/wiremock-user/AxH20Ea-m-U), but I figured I'd comment here too. I've been working on an admin API extension, https://github.com/MasonM/wiremock-snapshot, that can currently solve the first two objectives (snapshot recording and recording parameters). It's alpha and doesn't have feature-parity with Record & Playback yet. PRs and feedback appreciated.

**Comment by MasonM** [comment]
For those following this issue: I just entered PR #674 to integrate wiremock-snapshot into wiremock
